File: LeetCode-DSA/LeetCode75/Level2/HouseRobber.cpp
/*
Q198. Description: You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
*/

class Solution {
public:
    int rob(vector<int>& nums) 
    {
        if(nums.size() == 0)
            return 0;
        if(nums.size() == 1)
            return nums[0];
        
        vector<int> profit;
        profit.push_back(nums[0]);
        // Start with maximum 0th or 1st house
        profit.push_back(nums[0] > nums[1] ? nums[0] : nums[1]);
        
        // Skip adjacent houses
        for(int i = 2; i < nums.size(); i++)
            profit.push_back(max(profit[i-2] + nums[i], profit[i-1]));
        return profit[nums.size() - 1];
    }
};

File: LeetCode-DSA/Algorithms1/HouseRobber.py
'''
Q198. Description: You are a professional robber planning to rob houses along a street. Each house has a certain amount of money stashed, the only constraint stopping you from robbing each of them is that adjacent houses have security systems connected and it will automatically contact the police if two adjacent houses were broken into on the same night.

Given an integer array nums representing the amount of money of each house, return the maximum amount of money you can rob tonight without alerting the police.
'''

class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return max(nums)
        profit = {}
        # Index 0 holds even start point
        profit[0] = nums[0]
        # Index 1 stores maximum value at even or odd start point
        profit[1] = max(nums[0], nums[1])
        
        for i in range(2, len(nums)):
            # Store even+even and odd+odd sums
            profit[i] = nums[i] + profit[i-2]
            # Compare value with previous maximum
            profit[i] = max(profit[i], profit[i-1])
        # Largest value is stored at end of profit hashmap    
        return profit[len(profit)-1]